### ISSUE TITLE
Error: Cannot find module 'azure-pipelines-task-lib/task'

### ISSUE THREAD

**Issue by hoffiz** [Microsoft/azure-pipelines-task-lib#485]
I have an extension with multiple Azure pipeline tasks in it. Some of the task properly install the dependencies listed in the package.json but others do not. Looking at only 2 tasks, one that downloads the necessary modules into \node_modules and one that does not shows no possible suspect.

I can manually, for the tasks that don't download it, run npm install and my pipeline build will work properly and will no longer get the:

 Error: Cannot find module 'azure-pipelines-task-lib/task'

Any suggestions of what to try or review? 

Looking at similar submitted issues I could not find any clues of what is causing this bug on my side.

**Comment by bryanmacfarlane** [comment]
Tasks need to carry their dependencies.  In other words, node_modules and npm install is a task build time operation (not runtime).

> Some of the task properly install the dependencies listed in the package.json but others do not

A task shouldn't install at runtime.  Also, every task needs to carry it's full dependency tree (every task is self contained) even if there's multiple in one extension.

Does that help?

**Comment by hoffiz** [comment]
It does thanks!

**Comment by Farzad-Jalali** [comment]
> Tasks need to carry their dependencies. In other words, node_modules and npm install is a task build time operation (not runtime).
> 
> > Some of the task properly install the dependencies listed in the package.json but others do not
> 
> A task shouldn't install at runtime. Also, every task needs to carry it's full dependency tree (every task is self contained) even if there's multiple in one extension.
> 
> Does that help?

Could you be more specific about how to do it?
thanks in advance

**Comment by damccorm** [comment]
Basically, before you publish your task you should run npm install and then you should include the node modules in the package that you are publishing. Does that make sense? If not, could you be more specific about where the issue is?

**Comment by shudv** [comment]
For anyone who hits this going forward, you can use this [starter template](https://github.com/shudv/azure-pipelines-task-template) for quickly get up and running with a custom pipelines task.

**Comment by Mykyta** [comment]
Hi guys, I've faced with the same error. But to reduce the extension package size, I point the path to lib in manifest file, such as 
```
 {"path": "nexusIQ/node_modules/azure-pipelines-task-lib"},
 {"path": "nexusIQ/node_modules/balanced-match"},
{"path": "nexusIQ/node_modules/brace-expansion"},
{"path": "nexusIQ/node_modules/concat-map"},
 {"path": "nexusIQ/node_modules/minimatch"},
 {"path": "nexusIQ/node_modules/mockery"},
{"path": "nexusIQ/node_modules/q"},
{"path": "nexusIQ/node_modules/semver"},
 {"path": "nexusIQ/node_modules/shelljs"},
{"path": "nexusIQ/node_modules/uuid"}
```
with whole dependencies. Maybe some resolve that issue?

**Comment by CodeTroopers** [comment]
I have the same error. I tried with glob path expression in the manifest file with :
`"path": "./node_modules/azure-pipelines-task-lib/**/*.js"`
but creating package fails with `Error: ENOENT: no such file or directory, lstat 'C:\Dev\Extension\Task\node_modules\azure-pipelines-task-lib\**\*.js'`

What is the correct way to include dependencies in the extension? I also tried to compile TypeScript in a single file but commonjs module is not supported!

**Comment by mansellan** [comment]
I had this too - in my case it was because the `node_modules` directory was not in the same directory as the task script.

**Comment by tillig** [comment]
This makes it really hard to do any sort of shared code across extensions.

```
/Common
  common.ts
/Extension1
  task.ts # references common.ts
  task.json
/Extension2
  task.ts # references common.ts
  task.json
/node_modules # has modules used by all three
```

I see [in the official azure-pipelines-tasks repo](https://github.com/microsoft/azure-pipelines-tasks) that any shared code has a really crazy thing where the common code gets packed into a .tgz and then referenced as a .tgz from each task. I'm guessing that's why.

It means any shared code you might have needs to be published as a module outside the task repo and then it'll end up being included _multiple times_ if you have multiple extensions (or a multiple-version extension) in one package.

Is there really no workaround for having more shared code in a single package?

**Comment by tillig** [comment]
Related: I noticed that on a local dev box doing `node task.js` on a task can correctly invoke it and the standard Node module search works; it _very much doesn't work_ on the hosted build agents. That makes it really hard to debug or find these issues out. I think I've buried two days in trying to get a shared library to work, getting it totally working on my local machine, and then having it _totally not work_ in the actual agent.

**Comment by jessehouwing** [comment]
I used a slightly different approach here.

https://github.com/microsoft/azure-devops-extension-tasks

But every take must be self contained. Has to do with the fact that tasks pre-date the marketplace extensions and are managed individually internally.

**Comment by tillig** [comment]
@jessehouwing I was about to file a whole "can we improve the packaging situation" issue but I'm guessing there's nothing to be done there. Like, you can't use webpack because [the `loc()` localization function is hard-tied to the filesystem](https://github.com/microsoft/azure-pipelines-task-lib/blob/34e17f6ca02c256bcb4e121880f6d4b7da65eed2/node/internal.ts#L183) and if it can't find the files you end up in a stack overflow error. (Yeah, I tried). Which means you may also get pretty limited by the 50MB VSIX max package size if you have build tasks of sufficient complexity - just a Hello World level build task packs in a VSIX at about 500KB - 1MB.

Thanks for the example, I see how you've added things to the `files` in `tsconfig.json` instead of using `references`, which I'm guessing is part of the magic. I also see `package.json` in every folder. I'll see if I can update my extension to be a little closer to that. I had that, then had a _lot_ of challenge getting the common code references to work. Finally got it, then ended up here. The docs are definitely not sufficient when it comes to putting multiple build tasks in a single extension.

**Comment by jessehouwing** [comment]
With the direction GitHub actions are taking, only a single task per repo... Since the underlying infra is the same... I'm working to make new tasks with just a single task that has an option to select what functionality to trigger.

For really large tasks, create a bootstrapper task that runs npm install on first run. Or that downloads a payload from a binary blob in azure storage.

**Comment by tillig** [comment]
[Back up a few comments](https://github.com/microsoft/azure-pipelines-task-lib/issues/485#issuecomment-449187516) it appears we _shouldn't_ be running `npm install` on first run, but I see what you're saying. The GitHub Actions note is interesting and makes me think perhaps I should go that one-repo-per-task route for our internal actions, too, just so we don't have to refactor it all out later. It'd be more repos/VSIX packages right now, but we'd be ready for the switch when it's time.

**Comment by jessehouwing** [comment]
Yeah, so you can do a fully self-contained task with its own node_modules, but have it run npm install on a subfolder. Like how the Extension Tasks run npm install to install tfx-cli.
